Apps/NotSensorLogger2/app/src/main/java/com/example/notsensorlogger2/MainActivity.java

```diff
@@ -0,0 +1,353 @@
+     *
+     * @param requestCode  The request code passed in requestPermissions.
+     * @param permissions  The requested permissions. Never null.
+     * @param grantResults The grant results for the corresponding permissions which is either PERMISSION_GRANTED or PERMISSION_DENIED. Never null.
+     */
+    @Override
+    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
+        if (requestCode == PERMISSION_REQUEST_RECORD_AUDIO) {
+            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                if (soundLevelHandler != null) {
+                    soundLevelHandler.initAudioRecord();
+                }
+            }
+        }
+    }
+
+    /**
+     * Sets up sensor list and RecyclerView for sensor toggling, initializes sensor handlers, and checks permissions.
+     * Initializes and configures buttons for controlling data collection and navigating to other activities.
+     *
+     * @param savedInstanceState Contains data supplied in onSaveInstanceState(Bundle) or null if no data was supplied.
+     */
+    public final void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_main);
+
+        deviceId = DeviceIdManager.getDeviceId(this);
+        Log.d("DeviceId", "Device ID: " + deviceId);
+
+        sensorManager = (SensorManager) getSystemService(SENSOR_SERVICE);
+        accelerometerHandler = new AccelerometerHandler(sensorManager);
+        gyroscopeHandler = new GyroscopeHandler(sensorManager);
+        magnetometerHandler = new MagnetometerHandler(sensorManager);
+        lightSensorHandler = new LightSensorHandler(sensorManager);
+        humiditySensorHandler = new RelativeHumiditySensorHandler(sensorManager);
+        pressureSensorHandler = new PressureSensorHandler(sensorManager);
+        gravitySensorHandler = new GravitySensorHandler(sensorManager);
+        locationTracker = new LocationHandler((Context) this);
+        soundLevelHandler = new SoundLevelHandler(this, sensorManager);
+
+        // After initializing all sensors and handlers, call the setup method for RecyclerView
+        setupSensorList();
+
+        Button stopDataButton = findViewById(R.id.start_recording_button);
+        stopDataButton.setText(R.string.start_button_text);
+
+        isDataCollectionActive = false;
+
+        stopDataButton.setOnClickListener(v -> {
+            if (!isDataCollectionActive) {
+                startDataCollection();
+                stopDataButton.setText(R.string.stop_button_text);
+                isDataCollectionActive = true;
+            } else {
+                stopDataCollection();
+                stopDataButton.setText(R.string.start_button_text);
+                isDataCollectionActive = false;
```

Comment:
it is set in stopDataCollection() already

Reply:
Fixed